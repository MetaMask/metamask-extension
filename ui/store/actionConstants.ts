export const GO_HOME = 'GO_HOME';
// modal state
export const MODAL_OPEN = 'UI_MODAL_OPEN';
export const MODAL_CLOSE = 'UI_MODAL_CLOSE';
export const SET_CONFIRMATION_EXCHANGE_RATES =
  'SET_CONFIRMATION_EXCHANGE_RATES';
// alert state
export const ALERT_OPEN = 'UI_ALERT_OPEN';
export const ALERT_CLOSE = 'UI_ALERT_CLOSE';
export const QR_CODE_DETECTED = 'UI_QR_CODE_DETECTED';
// network dropdown open
export const NETWORK_DROPDOWN_OPEN = 'UI_NETWORK_DROPDOWN_OPEN';
export const NETWORK_DROPDOWN_CLOSE = 'UI_NETWORK_DROPDOWN_CLOSE';
export const IMPORT_NFTS_MODAL_OPEN = 'UI_IMPORT_NFTS_MODAL_OPEN';
export const IMPORT_NFTS_MODAL_CLOSE = 'UI_IMPORT_NFTS_MODAL_CLOSE';
export const SHOW_IPFS_MODAL_OPEN = 'UI_IPFS_MODAL_OPEN';
export const SHOW_PERMITTED_NETWORK_TOAST_OPEN =
  'UI_PERMITTED_NETWORK_TOAST_OPEN';
export const SHOW_PERMITTED_NETWORK_TOAST_CLOSE =
  'UI_PERMITTED_NETWORK_TOAST_CLOSE';
export const SHOW_IPFS_MODAL_CLOSE = 'UI_IPFS_MODAL_CLOSE';
export const IMPORT_TOKENS_POPOVER_OPEN = 'UI_IMPORT_TOKENS_POPOVER_OPEN';
export const IMPORT_TOKENS_POPOVER_CLOSE = 'UI_IMPORT_TOKENS_POPOVER_CLOSE';
export const SHOW_BASIC_FUNCTIONALITY_MODAL_OPEN =
  'SHOW_BASIC_FUNCTIONALITY_MODAL_OPEN';
export const SHOW_BASIC_FUNCTIONALITY_MODAL_CLOSE =
  'SHOW_BASIC_FUNCTIONALITY_MODAL_CLOSE';
export const ONBOARDING_TOGGLE_BASIC_FUNCTIONALITY_ON =
  'ONBOARDING_TOGGLE_BASIC_FUNCTIONALITY_ON';
export const ONBOARDING_TOGGLE_BASIC_FUNCTIONALITY_OFF =
  'ONBOARDING_TOGGLE_BASIC_FUNCTIONALITY_OFF';

// remote state
export const UPDATE_METAMASK_STATE = 'UPDATE_METAMASK_STATE';
export const SELECTED_ADDRESS_CHANGED = 'SELECTED_ADDRESS_CHANGED';
export const SELECTED_ACCOUNT_CHANGED = 'SELECTED_ACCOUNT_CHANGED';
export const ACCOUNT_CHANGED = 'ACCOUNT_CHANGED';
export const CHAIN_CHANGED = 'CHAIN_CHANGED';
export const ADDRESS_BOOK_UPDATED = 'ADDRESS_BOOK_UPDATED';
export const GAS_FEE_ESTIMATES_UPDATED = 'GAS_FEE_ESTIMATES_UPDATED';
export const CLOSE_WELCOME_SCREEN = 'CLOSE_WELCOME_SCREEN';
// send state
export const CLEAR_SWAP_AND_SEND_STATE = 'CLEAR_SWAP_AND_SEND_STATE';
// unlock screen
export const UNLOCK_IN_PROGRESS = 'UNLOCK_IN_PROGRESS';
export const UNLOCK_FAILED = 'UNLOCK_FAILED';
export const UNLOCK_SUCCEEDED = 'UNLOCK_SUCCEEDED';
export const LOCK_METAMASK = 'LOCK_METAMASK';
// error handling
export const DISPLAY_WARNING = 'DISPLAY_WARNING';
export const HIDE_WARNING = 'HIDE_WARNING';
export const SHOW_SETTINGS_PAGE_ERROR = 'SHOW_SETTINGS_PAGE_ERROR';
export const HIDE_SETTINGS_PAGE_ERROR = 'HIDE_SETTINGS_PAGE_ERROR';
export const CAPTURE_SINGLE_EXCEPTION = 'CAPTURE_SINGLE_EXCEPTION';
// accounts screen
export const SHOW_ACCOUNTS_PAGE = 'SHOW_ACCOUNTS_PAGE';
export const SHOW_CONF_TX_PAGE = 'SHOW_CONF_TX_PAGE';
// account detail screen
export const SHOW_SEND_TOKEN_PAGE = 'SHOW_SEND_TOKEN_PAGE';
export const SHOW_PRIVATE_KEY = 'SHOW_PRIVATE_KEY';
export const SET_ACCOUNT_LABEL = 'SET_ACCOUNT_LABEL';
export const CLEAR_ACCOUNT_DETAILS = 'CLEAR_ACCOUNT_DETAILS';
export const SET_ACCOUNT_DETAILS_ADDRESS = 'SET_ACCOUNT_DETAILS_ADDRESS';
// tx conf screen
export const COMPLETED_TX = 'COMPLETED_TX';
export const TRANSACTION_ERROR = 'TRANSACTION_ERROR';
export const UPDATE_TRANSACTION_PARAMS = 'UPDATE_TRANSACTION_PARAMS';
export const SET_NEXT_NONCE = 'SET_NEXT_NONCE';
// config screen
export const SET_HARDWARE_WALLET_DEFAULT_HD_PATH =
  'SET_HARDWARE_WALLET_DEFAULT_HD_PATH';
// loading overlay
export const SHOW_LOADING = 'SHOW_LOADING_INDICATION';
export const HIDE_LOADING = 'HIDE_LOADING_INDICATION';

// Nft still fetching indication spinners
export const SHOW_NFT_STILL_FETCHING_INDICATION =
  'SHOW_NFT_STILL_FETCHING_INDICATION';
export const HIDE_NFT_STILL_FETCHING_INDICATION =
  'HIDE_NFT_STILL_FETCHING_INDICATION';

export const SHOW_NFT_DETECTION_ENABLEMENT_TOAST =
  'SHOW_NFT_DETECTION_ENABLEMENT_TOAST';

export const TOGGLE_ACCOUNT_MENU = 'TOGGLE_ACCOUNT_MENU';
export const TOGGLE_NETWORK_MENU = 'TOGGLE_NETWORK_MENU';
export const SET_SELECTED_ACCOUNTS_FOR_DAPP_CONNECTIONS =
  'SET_SELECTED_ACCOUNTS_FOR_DAPP_CONNECTIONS';
export const SET_SELECTED_NETWORKS_FOR_DAPP_CONNECTIONS =
  'SET_SELECTED_NETWORKS_FOR_DAPP_CONNECTIONS';

// deprecated network modal
export const DEPRECATED_NETWORK_POPOVER_OPEN =
  'DEPRECATED_NETWORK_POPOVER_OPEN';
export const DEPRECATED_NETWORK_POPOVER_CLOSE =
  'DEPRECATED_NETWORK_POPOVER_CLOSE';

// preferences
export const UPDATE_CUSTOM_NONCE = 'UPDATE_CUSTOM_NONCE';

export const SET_PARTICIPATE_IN_METAMETRICS = 'SET_PARTICIPATE_IN_METAMETRICS';
export const SET_DATA_COLLECTION_FOR_MARKETING =
  'SET_DATA_COLLECTION_FOR_MARKETING';
export const DELETE_METAMETRICS_DATA_MODAL_OPEN =
  'DELETE_METAMETRICS_DATA_MODAL_OPEN';
export const DELETE_METAMETRICS_DATA_MODAL_CLOSE =
  'DELETE_METAMETRICS_DATA_MODAL_CLOSE';
export const DATA_DELETION_ERROR_MODAL_OPEN =
  'DELETE_METAMETRICS_DATA_ERROR_MODAL_OPEN';
export const DATA_DELETION_ERROR_MODAL_CLOSE =
  'DELETE_METAMETRICS_DATA_ERROR_MODAL_CLOSE';

// locale
export const SET_CURRENT_LOCALE = 'SET_CURRENT_LOCALE';

// Onboarding

export const COMPLETE_ONBOARDING = 'COMPLETE_ONBOARDING';
export const RESET_ONBOARDING = 'RESET_ONBOARDING';
export const ONBOARDED_IN_THIS_UI_SESSION = 'ONBOARDED_IN_THIS_UI_SESSION';

// Ledger

export const SET_WEBHID_CONNECTED_STATUS = 'SET_WEBHID_CONNECTED_STATUS';
export const SET_LEDGER_TRANSPORT_STATUS = 'SET_LEDGER_TRANSPORT_STATUS';

// Network
export const SET_PENDING_TOKENS = 'SET_PENDING_TOKENS';
export const CLEAR_PENDING_TOKENS = 'CLEAR_PENDING_TOKENS';

export const SET_FIRST_TIME_FLOW_TYPE = 'SET_FIRST_TIME_FLOW_TYPE';

export const SET_SELECTED_NETWORK_CONFIGURATION_ID =
  'SET_SELECTED_NETWORK_CONFIGURATION_ID';
export const SET_NEW_NETWORK_ADDED = 'SET_NEW_NETWORK_ADDED';
export const SET_EDIT_NETWORK = 'SET_EDIT_NETWORK';

export const SET_NEW_NFT_ADDED_MESSAGE = 'SET_NEW_NFT_ADDED_MESSAGE';
export const SET_REMOVE_NFT_MESSAGE = 'SET_REMOVE_NFT_MESSAGE';

export const LOADING_METHOD_DATA_STARTED = 'LOADING_METHOD_DATA_STARTED';
export const LOADING_METHOD_DATA_FINISHED = 'LOADING_METHOD_DATA_FINISHED';

export const SET_REQUEST_ACCOUNT_TABS = 'SET_REQUEST_ACCOUNT_TABS';
export const SET_OPEN_METAMASK_TAB_IDS = 'SET_OPEN_METAMASK_TAB_IDS';

// Home Screen
export const HIDE_WHATS_NEW_POPUP = 'HIDE_WHATS_NEW_POPUP';
export const TOGGLE_GAS_LOADING_ANIMATION = 'TOGGLE_GAS_LOADING_ANIMATION';

// Smart Transactions
export const SET_SMART_TRANSACTIONS_ERROR = 'SET_SMART_TRANSACTIONS_ERROR';
export const DISMISS_SMART_TRANSACTIONS_ERROR_MESSAGE =
  'DISMISS_SMART_TRANSACTIONS_ERROR_MESSAGE';

export const TOGGLE_CURRENCY_INPUT_SWITCH = 'TOGGLE_CURRENCY_INPUT_SWITCH';

// Token detection v2
export const SET_NEW_TOKENS_IMPORTED = 'SET_NEW_TOKENS_IMPORTED';

export const SET_NEW_TOKENS_IMPORTED_ERROR = 'SET_NEW_TOKENS_IMPORTED_ERROR';

// Token allowance
export const SET_CUSTOM_TOKEN_AMOUNT = 'SET_CUSTOM_TOKEN_AMOUNT';

// Connections Modal
export const CONNECT_ACCOUNTS_MODAL_OPEN = 'UI_CONNECT_ACCOUNTS_MODAL_OPEN';
export const CONNECT_ACCOUNTS_MODAL_CLOSE = 'UI_CONNECT_ACCOUNTS_MODAL_CLOSE';

///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
export const SHOW_KEYRING_SNAP_REMOVAL_RESULT =
  'SHOW_KEYRING_SNAP_REMOVAL_RESULT';
export const HIDE_KEYRING_SNAP_REMOVAL_RESULT =
  'HIDE_KEYRING_SNAP_REMOVAL_RESULT';
///: END:ONLY_INCLUDE_IF

export const SET_SHOW_NFT_AUTO_DETECT_MODAL_UPGRADE =
  'SET_SHOW_NFT_AUTO_DETECT_MODAL_UPGRADE';

export const TOKEN_SORT_CRITERIA = 'TOKEN_SORT_CRITERIA';
