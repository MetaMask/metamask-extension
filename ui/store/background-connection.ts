import pify from 'pify';

let background:
  | ({
      connectionStream: { readable: boolean };
      DisconnectError: typeof Error;
      // TODO: Replace `any` with type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } & Record<string, (...args: any[]) => any>)
  | null = null;
let promisifiedBackground: Record<
  string,
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (...args: any[]) => Promise<any>
> | null = null;

export const generateActionId = () => Date.now() + Math.random();

/**
 * Promise-style call to background method invokes promisifiedBackground method directly.
 *
 * @param method - name of the background method
 * @param [args] - arguments to that method, if any
 * @returns
 */
export function submitRequestToBackground<R>(
  method: string,
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  args?: any[],
): Promise<R> {
  return promisifiedBackground?.[method](
    ...(args ?? []),
  ) as unknown as Promise<R>;
}

type CallbackMethod<R = unknown> = (error?: unknown, result?: R) => void;

/**
 * [Deprecated] Callback-style call to background method
 * invokes promisifiedBackground method directly.
 *
 * @deprecated Use async `submitRequestToBackground` function instead.
 * @param method - name of the background method
 * @param [args] - arguments to that method, if any
 * @param callback - Node style (error, result) callback for finishing the operation
 */
export const callBackgroundMethod = <R>(
  method: string,
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  args: any[],
  callback: CallbackMethod<R>,
) => {
  background?.[method](...args, callback);
};

/**
 * Sets/replaces the background connection reference
 * Under MV3 it also triggers queue processing if the new background is connected
 *
 * @param backgroundConnection
 */
export async function setBackgroundConnection(
  backgroundConnection: typeof background,
) {
  background = backgroundConnection;
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  promisifiedBackground = pify(background as Record<string, any>);
}
