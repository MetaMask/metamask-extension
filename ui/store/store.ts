import { StoreEnhancer } from 'redux';
import { configureStore as baseConfigureStore } from '@reduxjs/toolkit';
import devtoolsEnhancer from 'remote-redux-devtools';
import { ApprovalControllerState } from '@metamask/approval-controller';
import { GasEstimateType, GasFeeEstimates } from '@metamask/gas-fee-controller';
import { TransactionMeta } from '@metamask/transaction-controller';
import { InternalAccount } from '@metamask/keyring-api';
import {
  NftControllerState,
  TokensControllerState,
} from '@metamask/assets-controllers';
import { NotificationServicesControllerState } from '@metamask/notification-services-controller/notification-services';
import rootReducer from '../ducks';
import { LedgerTransportTypes } from '../../shared/constants/hardware-wallets';
import type { NetworkStatus } from '../../shared/constants/network';

/**
 * This interface is temporary and is copied from the message-manager.js file
 * and is the 'msgParams' key of the interface declared there. We should get a
 * universal Message type to use for this, the Message manager and all
 * the other types of messages.
 *
 * TODO: Replace this
 */
export type TemporaryMessageDataType = {
  id: string;
  type: string;
  msgParams: {
    metamaskId: string;
    data: string;
  };
  ///: BEGIN:ONLY_INCLUDE_IF(build-mmi)
  metadata?: {
    custodyId?: string;
  };
  status?: string;
  ///: END:ONLY_INCLUDE_IF
};

export type MessagesIndexedById = {
  [id: string]: TemporaryMessageDataType;
};

/**
 * This interface is a temporary interface to describe the state tree that is
 * sent from the background. Ideally we can build this using Types in the
 * backend when we compose the stores, then we can import it here and use it.
 *
 * Some of this is duplicated in the metamask redux duck. In *most* cases the
 * state received from the background takes precedence over anything in the
 * metamask reducer.
 */
type TemporaryBackgroundState = NftControllerState &
  NotificationServicesControllerState &
  TokensControllerState & {
    addressBook: {
      [chainId: string]: {
        name: string;
      }[];
    };
    // todo: can this be deleted post network controller v20
    providerConfig: {
      chainId: string;
    };
    transactions: TransactionMeta[];
    ledgerTransportType: LedgerTransportTypes;
    unapprovedDecryptMsgs: MessagesIndexedById;
    unapprovedPersonalMsgs: MessagesIndexedById;
    unapprovedTypedMessages: MessagesIndexedById;
    networksMetadata: {
      [NetworkClientId: string]: {
        EIPS: { [eip: string]: boolean };
        status: NetworkStatus;
      };
    };
    selectedNetworkClientId: string;
    pendingApprovals: ApprovalControllerState['pendingApprovals'];
    approvalFlows: ApprovalControllerState['approvalFlows'];
    knownMethodData?: {
      [fourBytePrefix: string]: Record<string, unknown>;
    };
    gasFeeEstimates: GasFeeEstimates;
    gasEstimateType: GasEstimateType;
    ///: BEGIN:ONLY_INCLUDE_IF(build-mmi)
    // TODO: Replace `any` with type
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    custodyAccountDetails?: { [key: string]: any };
    ///: END:ONLY_INCLUDE_IF
    internalAccounts: {
      accounts: {
        [key: string]: InternalAccount;
      };
      selectedAccount: string;
    };
    keyrings: { type: string; accounts: string[] }[];
  };

type RootReducerReturnType = ReturnType<typeof rootReducer>;

export type CombinedBackgroundAndReduxState = RootReducerReturnType & {
  activeTab: {
    origin: string;
  };
  metamask: RootReducerReturnType['metamask'] & TemporaryBackgroundState;
  appState: RootReducerReturnType['appState'];
  send: RootReducerReturnType['send'];
  DNS: RootReducerReturnType['DNS'];
  history: RootReducerReturnType['history'];
  confirmAlerts: RootReducerReturnType['confirmAlerts'];
  confirmTransaction: RootReducerReturnType['confirmTransaction'];
  swaps: RootReducerReturnType['swaps'];
  bridge: RootReducerReturnType['bridge'];
  gas: RootReducerReturnType['gas'];
  localeMessages: RootReducerReturnType['localeMessages'];
};

// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default function configureStore(preloadedState: any) {
  const debugModeEnabled = Boolean(process.env.METAMASK_DEBUG);
  const isDev = debugModeEnabled && !process.env.IN_TEST;
  const enhancers: StoreEnhancer[] = [];

  if (isDev) {
    enhancers.push(
      devtoolsEnhancer({
        name: 'MetaMask',
        hostname: 'localhost',
        port: 8000,
        realtime: true,
      }) as StoreEnhancer,
    );
  }

  return baseConfigureStore({
    reducer: rootReducer as () => CombinedBackgroundAndReduxState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        /**
         * We do not persist the redux tree for rehydration, so checking for
         * serializable state keys is not relevant for now. Any state that persists
         * is managed in the background. We may at some point want this, but we can
         * gradually implement by using the ignore options to ignore those actions
         * and state keys that are not serializable, preventing us from adding new
         * actions and state that would violate our ability to persist state keys.
         * NOTE: redux-thunk is included by default in the middleware below.
         */
        serializableCheck: false,
        /**
         * immutableCheck controls whether we get warnings about mutation of
         * state, this is turned off by default for now since it heavily affects
         * performance due to the Redux state growing larger.
         */
        immutableCheck: false,
      }),
    devTools: false,
    enhancers,
    preloadedState,
  });
}
type Store = ReturnType<typeof configureStore>;
export type MetaMaskReduxState = ReturnType<Store['getState']>;
export type MetaMaskReduxDispatch = Store['dispatch'];
