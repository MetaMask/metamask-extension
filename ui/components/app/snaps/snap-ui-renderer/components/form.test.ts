import {
  Box,
  Form,
  Input,
  Button,
  Field,
  Checkbox,
} from '@metamask/snaps-sdk/jsx';
import { fireEvent } from '@testing-library/react';
import * as backgroundConnection from '../../../../../store/background-connection';
import {
  MOCK_INTERFACE_ID,
  MOCK_SNAP_ID,
  renderInterface,
} from '../test-utils';

jest.mock('../../../../../store/background-connection', () => ({
  ...jest.requireActual('../../../../../store/background-connection'),
  submitRequestToBackground: jest.fn(),
}));

const { submitRequestToBackground } = jest.mocked(backgroundConnection);

describe('SnapUIForm', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('will render', () => {
    const { container, getByRole } = renderInterface(
      Box({
        children: Form({
          name: 'form',
          children: [
            Input({ name: 'input' }),
            Button({ type: 'submit', name: 'submit', children: 'Submit' }),
          ],
        }),
      }),
    );

    const input = getByRole('textbox');
    fireEvent.change(input, { target: { value: 'abc' } });

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      1,
      'updateInterfaceState',
      [MOCK_INTERFACE_ID, { form: { input: 'abc' } }],
    );

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      2,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: { name: 'input', type: 'InputChangeEvent', value: 'abc' },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    const button = getByRole('button');
    fireEvent.click(button);

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      3,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: { name: 'submit', type: 'ButtonClickEvent' },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      4,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: {
                name: 'form',
                type: 'FormSubmitEvent',
                value: {
                  input: 'abc',
                },
              },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    expect(container).toMatchSnapshot();
  });

  it('will render with fields', () => {
    const { container, getByRole } = renderInterface(
      Box({
        children: Form({
          name: 'form',
          children: [
            Field({ label: 'My Input', children: Input({ name: 'input' }) }),
            Field({
              label: 'Checkbox',
              children: Checkbox({ name: 'checkbox' }),
            }),
            Button({ type: 'submit', name: 'submit', children: 'Submit' }),
          ],
        }),
      }),
    );

    const input = getByRole('textbox');
    fireEvent.change(input, { target: { value: 'abc' } });

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      1,
      'updateInterfaceState',
      [MOCK_INTERFACE_ID, { form: { input: 'abc' } }],
    );

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      2,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: { name: 'input', type: 'InputChangeEvent', value: 'abc' },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    const checkbox = getByRole('checkbox');
    fireEvent.click(checkbox);

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      3,
      'updateInterfaceState',
      [MOCK_INTERFACE_ID, { form: { checkbox: true, input: 'abc' } }],
    );

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      4,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: {
                name: 'checkbox',
                type: 'InputChangeEvent',
                value: true,
              },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    const button = getByRole('button');
    fireEvent.click(button);

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      5,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: { name: 'submit', type: 'ButtonClickEvent' },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    expect(submitRequestToBackground).toHaveBeenNthCalledWith(
      6,
      'handleSnapRequest',
      [
        {
          handler: 'onUserInput',
          origin: 'metamask',
          request: {
            jsonrpc: '2.0',
            method: ' ',
            params: {
              event: {
                name: 'form',
                type: 'FormSubmitEvent',
                value: {
                  checkbox: true,
                  input: 'abc',
                },
              },
              id: MOCK_INTERFACE_ID,
            },
          },
          snapId: MOCK_SNAP_ID,
        },
      ],
    );

    expect(container).toMatchSnapshot();
  });
});
