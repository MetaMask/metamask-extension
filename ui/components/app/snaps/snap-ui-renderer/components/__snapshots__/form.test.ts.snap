// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapUIForm will render 1`] = `
<div>
  <div
    class="mm-box snap-ui-renderer__content mm-box--height-full mm-box--background-color-background-alternative"
    style="overflow-y: auto; margin-bottom: 0px;"
  >
    <div
      class="box snap-ui-renderer__container box--display-flex box--flex-direction-column"
    >
      <div
        class="box snap-ui-renderer__panel box--display-flex box--flex-direction-column box--justify-content-flex-start box--color-text-default"
      >
        <form
          class="mm-box snap-ui-renderer__form mm-box--display-flex mm-box--gap-2 mm-box--flex-direction-column"
          id="form"
        >
          <div
            class="mm-box mm-form-text-field snap-ui-renderer__input mm-box--display-flex mm-box--flex-direction-column"
          >
            <div
              class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
            >
              <input
                autocomplete="off"
                class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-md mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
                focused="false"
                id="input"
                type="text"
                value="abc"
              />
            </div>
          </div>
          <button
            class="mm-box mm-text snap-ui-renderer__button mm-text--body-md mm-text--font-weight-medium mm-box--color-info-default"
            id="submit"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
    <span
      data-renders="1"
      data-testid="performance"
    />
  </div>
</div>
`;

exports[`SnapUIForm will render with fields 1`] = `
<div>
  <div
    class="mm-box snap-ui-renderer__content mm-box--height-full mm-box--background-color-background-alternative"
    style="overflow-y: auto; margin-bottom: 0px;"
  >
    <div
      class="box snap-ui-renderer__container box--display-flex box--flex-direction-column"
    >
      <div
        class="box snap-ui-renderer__panel box--display-flex box--flex-direction-column box--justify-content-flex-start box--color-text-default"
      >
        <form
          class="mm-box snap-ui-renderer__form mm-box--display-flex mm-box--gap-2 mm-box--flex-direction-column"
          id="form"
        >
          <div
            class="mm-box mm-form-text-field snap-ui-renderer__input snap-ui-renderer__field mm-box--display-flex mm-box--flex-direction-column"
          >
            <label
              class="mm-box mm-text mm-label mm-label--html-for mm-form-text-field__label mm-text--body-md mm-text--font-weight-medium mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
              for="input"
            >
              My Input
            </label>
            <div
              class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
            >
              <input
                autocomplete="off"
                class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-md mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
                focused="false"
                id="input"
                type="text"
                value="abc"
              />
            </div>
          </div>
          <div
            class="mm-box snap-ui-renderer__checkbox mm-box--display-flex mm-box--flex-direction-column"
          >
            <label
              class="mm-box mm-text mm-label mm-label--html-for mm-text--body-md mm-text--font-weight-medium mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
              for="checkbox"
            >
              Checkbox
            </label>
            <label
              class="mm-box mm-text mm-checkbox mm-text--body-md mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
            >
              <span
                class="mm-checkbox__input-wrapper"
              >
                <input
                  class="mm-box mm-checkbox__input mm-checkbox__input--checked mm-box--margin-0 mm-box--margin-right-0 mm-box--display-flex mm-box--background-color-primary-default mm-box--rounded-sm mm-box--border-color-border-muted mm-box--border-width-2 box--border-style-solid"
                  type="checkbox"
                />
                <span
                  class="mm-box mm-checkbox__icon mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-primary-inverse"
                  style="mask-image: url('./images/icons/check-bold.svg');"
                />
              </span>
            </label>
          </div>
          <button
            class="mm-box mm-text snap-ui-renderer__button mm-text--body-md mm-text--font-weight-medium mm-box--color-info-default"
            id="submit"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
    <span
      data-renders="1"
      data-testid="performance"
    />
  </div>
</div>
`;
